### ISSUE TITLE
PyInstaller doesn't import libraries requested by other library

### ISSUE THREAD

**Issue by padremayi** [pyinstaller/pyinstaller#2530]
PyInstaller doesn't import libraries requested by other library.

If I run my code with
`python3 my_code.py`
everything is ok, but, if I previously packaged my program, when I launch the executable some libraries are missing. If I re-install these libraries with
`pip3 library install`
the library is packed correctly

**Comment by htgoebel** [comment]
Please have a look at [Make sure everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly) for tips to find out what is missing. Then please refer to the manual for how to solve this.

If you need further assistance, please provide more details, see [How to report Bugs](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs). Esp. please provide a *minimal* example to reproduce the error.

**Comment by padremayi** [comment]
Hi, thanks for your reply.

I'm using PyInstaller 3.2.1 and my software is splitted on different .py files (and library).

In my code the package [requests](http://docs.python-requests.org/en/master/) is imported and this package is correctly installed in my system. As I wrote before, if I launch my code with
`python3 Launcher.py`
everything is ok, not with the packaged file.

In the file warnLauncher.txt I see a lot of importing error:

> missing module named xxx

The problem is foocused on library needed by requests like urllib3.

During packaging I see:

> Loading module hook"hook-requests.py"...

**Comment by htgoebel** [comment]
To repeat myself: Follow *all* the instructions at [Make sure everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly) and [How to report Bugs](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs). 

Otherwise nobody will be able to help you.

**Comment by padremayi** [comment]
Sorry but I indicated PyInstaller version, Python version and the package that needs to import other libraries and for which the libraries aren't imported (requests that, for example, needs urllib3). I can't provide you more information, for sure not my source code.

If the code runs without packaging maybe the problem is in the packaging process, but the operation doesn't provides me any information/error.

I also start with a clean installation of PyInstaller but nothing changed.

I don't know what more info I can provide to you.

Thanks

**Comment by padremayi** [comment]
Ok, maybe all developers are not interested.

There are 2 methods to avoid this:
1. Reinstall the not-imported package with pip
2. Use hidden import method 

I don't know why the first one fixes the problem and why it is necessary the second... Maybe PyInstaller doesn't see second level of import?

I hope this help someone

**Comment by xoviat** [comment]
> Reinstall the not-imported package with pip

Pyinstaller does not support packages that are not installed with pip or anaconda.

**Comment by htgoebel** [comment]
@padremayi You obviously did not follow the instructions, which include "try the current development release". Your problem should be fixed there already, see fd018208c. Stop wasting other peoples time and do your homework.

> The problem is focused on library needed by requests like urllib3.

"Focussed on" is much to vague for finding a solution.

@xoviat 
> Pyinstaller does not support packages that are not installed with pip or anaconda.

This is not quite true :-) PyInstaller includes any package as long as it is to be found in PYTHONPATH. Only if the packages requires some meta-data to be frozen correctly, it needs to be installed using the standard. Any well, for quite some packages, hooks are required.

**Comment by padremayi** [comment]
> Stop wasting other peoples time and do your homework

You're a lord

**Comment by hykavitha** [comment]
@padremayi : did the issue got resolved? Were you able to run the executable/binary file in a non python environment without pyinstaller complaining about import errors?

**Comment by sdresche17** [comment]
I've tried the suggestions and I'm getting the urllib3 issue with pyinstaller. 

I'm trying to use the smartsheet sdk, which relies on urllib3 and without it my objects are just treated as strings

Inside my build warning file it says I'm missing my requests.packages.urllib3. Thus far I've tried the following:

1. upgrade requests
2. upgrade systemtools
3. uninstall and reinstall pyinstaller
4. put the import to urllib3 directly in my python script
5. add the package location to the pyinstall using --paths

Any suggestions? I apologize if this isn't the right forum, I'm new to posts on github.

**Comment by hykavitha** [comment]
@sdresche17 : I was finally able to get pyinstaller working
Initially I came across all kind of problems like importing and package issue. Unfortunately there are not a lot of good examples on pyinstaller. 
Lessons: 
1) pip, pyinstaller, python were not talking to the same python version (so I unanimously made everything python3)
2) missing packages 
3) missing path (importing code from sub folders, I had used sys_path and all and it dint work with pyinstaller).
 
problem 1 : pyinstaller --onefile file.py -rw-rw-r-- 1 
step 2: trying to run the binary file , as you know it fails. 
>dist$ ./file
Traceback (most recent call last):
  File "file.py", line 21, in <module>
ImportError: No module named 'google'
[9736] Failed to execute script file

So its missing some modules, lets install 
pip install grpcio
pip install grpcio-tools
Above dint help. it needs to come from setup.py (magic)

I realized spending many hours on this. The  solution is is two step: 
-need to get the path and packages inside python's dist_packages using setup.py to build and install. 
In my case it was some paths were missing, I used setup.py which have some required packages installation,  build and install setup.py, Which will get all the required packages inside your python's dist_packages. 

>python3 setup.py build
>python3 setup.py install

by running above two commands, your python3 will get all pkg_resources

problem 2: if you are importing some sub folders in a big project like mine, then its a path problem 
>  pyinstaller --hidden-import=pkg_resources --onefile file.py --path=path_for_my_import_folder rw-rw-r--
This path should be coming from python3's dist pacakges like this

pyinstaller --hidden-import=pkg_resources --onefile file.py --path=--path=/usr/local/lib/python3.5/dist-packages/snap_plugin_lib_py-0+untagged.8.gd264bd5.dirty-py3.5.egg/snap_plugin
 -rw-rw-r--
